### PR TITLE
Call breakcheck less often when adding completions

### PR DESCRIPTION
**Pull request by puremourning**
There are many different ways to add completions to the popup menu, some
of which are slow (such as scanning the filesystem). However, a number
are very fast loops, such as the vimscript complete() function and
various ways to add completions from a predefined list or dict.

In the fast cases, the calls to breakcheck can start to dominate
performance as it can be quite slow to call select() on some platforms.

Solution: Call it less often in these cases.

---

I used a script to do some profiling. Script essentially:

1. displays a popup. I wanted to see if the presence of popups impacted performance.
2. sets up a `completefunc` which:
* starts a 1 ms timer (because I wanted to mimic the behaviour of my actual plugin)
* the timer callback calls `complete()` with a list of 120 items, all of 4 characters in length

3 the script runs `feedkeys( "\<C-x>\<C-u>", 'xt!' )` in a loop and uses a `CompleteDone` autocmd to do `feedkeys( "\<Esc>" )`. Again this is not minimal, but I wanted to investigate a pattern similar to how a real async completion plugin works.

It does this loop 30,000 times.

Without this patch, the test takes (on my machine(TM)) 45 seconds, and after this patch, it takes 32 seconds.

---

Hottest path *without* this change, is mch_breakcheck (again, on My-machine(TM)):

Screenshot 2021-04-03 at 17 27 17

Screenshot 2021-04-03 at 17 27 51

*With* this change, the hottest path looks more relavant:

Screenshot 2021-04-03 at 17 28 30
 
And completes more quickly

Screenshot 2021-04-03 at 17 28 37

---

NOTE: These are not very scientific measurements, but the results are easily repeatable at least that mch_breakcheck is expensive when running in a tight loop.

Builds were done on macOS 11.3.2 with `-g -O2` using master as of today.

---

I've not extensively tested this, but it seems fairly safe. Let me know what you think.

### CODE REVIEW

**Comment by puremourning** [comment]
The profiling script:

```viml

" The code is taken directly from the example in the Vim documentation
function! s:FindStart() abort
  " locate the start of the word
  let line = getline('.')
  let start = col('.') - 1
  while start > 0 && line[start - 1] =~ '\a'
    let start -= 1
  endwhile
  return start
endfunction

let s:completions = []
for i in range( 10 )
  let s:completions += map(
        \ split('Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec'),
        \ 'v:val . i' )
endfor

" The code is taken directly from the example in the Vim documentation
function! s:CompleteMonth( base ) abort
  " find months matching with "a:base"
  return filter(copy(s:completions), "v:val =~ '^' . a:base")
endfunction

let s:complete_timer = -1

function s:KillTimer() abort
  call timer_stop( s:complete_timer )
  let s:complete_timer = -1
  augroup ATestClear
    au!
  augroup END
endfunction

function! s:DoAsyncCompletion( start_col, base, id ) abort
  call complete( a:start_col, s:CompleteMonth( a:base ) )
endfunction

" See :help complete-functions
function! CompleteAsync( findstart, base ) abort
  if a:findstart
    " We will work out the start position later
    return s:FindStart()
  endif

  " Kill any existing request
  call s:KillTimer()

  " Kill the timer when leaving insert mode
  augroup ATestClear
    au InsertLeave * ++once call <SID>KillTimer()
  augroup END

  " Do something complicated that takes time. Pass the current column (actually
  " the start column) and the 'query' (a:base) to the callback using a partial.
  let s:complete_timer =  timer_start( 1,
                                     \ function( "s:DoAsyncCompletion",
                                               \ [ col( '.' ), a:base ] ) )

  return v:none
endfunction

set completefunc=CompleteAsync
set shortmess+=c

let g:attest_async_delay = 1

function ExitInsertMode()
  call feedkeys( "\<Esc>" )
endfunction

augroup Test_Async
  au CompleteDone * call ExitInsertMode()
augroup END

call popup_create( [ 'This is some text', 'and some more' ], {} )


let i = 0
while i < 30000
  call feedkeys( "o\<C-x>\<C-u>", 'xt!' )

  let i += 1
endwhile

augroup Test_Async
  au!
augroup END

call popup_clear()

delfunc! ExitInsertMode
%bwipe!
qa
```

**Comment by puremourning** [comment]
Just to add one more piece, of the original run, 9s  (31%) of the runtime is used by mch_breakcheck:

Screenshot 2021-04-03 at 17 34 35

With this change mch_breakcheck is just 1.3%.

Screenshot 2021-04-03 at 17 36 00

File: src/insexpand.c [hunk 1]
```diff
@@ -681,6 +680,51 @@ ins_compl_add(
...
+
+/*
+ * Add a match to the list of matches, when adding them in a very fast tight
+ * loop, such as f_complete().
+ * If the given string is already in the list of completions, then return
+ * NOTDONE, otherwise add it to the list and return OK.  If there is an error,
+ * maybe because alloc() returns NULL, then FAIL is returned.
+ */
+    static int
+ins_compl_add_fast(
```
Comment (bfredl) [inline src/insexpand.c:713]: Instead of duplicating the function, why not add `int fast` as another parameter?

Comment (puremourning) [reply src/insexpand.c:713]: Seemed more explicit. Less branchy fewer args the stack. Habit I guess.

**Comment by bfredl** [comment]
did https://github.com/vim/vim/commit/440cf096fad7bf628974abc344343b823d79a006 really fix this? I don't see where `CP_FAST` is actually used.

**Comment by brammool** [comment]
It's used in three places, how can you not see that?

**Comment by puremourning** [comment]
I think Bjorne is suggesting that we need something like this:

```cpp
    if ( flags & CP_FAST )
	fast_breakcheck();
    else
	ui_breakcheck();
```

**Comment by bfredl** [comment]
@brammool it is being _set_ in three places but I don't see where it is being checked, like something like @puremourning  suggested.

**Comment by brammool** [comment]
> @brammool it is being _set_ in three places but I don't see where it
> is being checked, like something like @puremourning  suggested.

Ah, now I see what you mean.  Yes, that part somehow went missing.

-- 
hundred-and-one symptoms of being an internet addict:
55. You ask your doctor to implant a gig in your brain.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Pull request closed without merging**